<div class='alert-block alert-info'>
    <br>
    <h1 align="center"><b>  AI Project A4:</b> Biscuit Manufactoring </h1>
    <h3 align="center">AI Algorithms - DIA 2</h3>
    <h3 align="center">Oscar Pastural - Capucine Boudin - Richard Goudelin</a></h3>
    <br>
</div>

<div style="background: linear-gradient(to right, #f9f9f9, #ffffff); border: 1px solid #e2e2e2; padding: 20px; border-radius: 10px; margin: 20px; box-shadow: 0 6px 12px rgba(0, 0, 0, 0.05);">
    <h2 style="color: #2a85d0; text-align: center; padding-bottom: 10px; border-bottom: 4px solid #2a85d0; margin-bottom: 20px; font-family: 'Arial', sans-serif; letter-spacing: 1px; font-size: 24px; text-shadow: 1px 1px 2px #e2e2e2;">Table of Content (Clickable)</h2>
    <ul style="list-style-type: none; padding-left: 0; font-size: 16px; font-family: 'Arial', sans-serif; color: #333;">
        <li style="margin-bottom: 10px;"><a href="#1" style="text-decoration: none; color: #2a85d0; transition: color 0.3s ease; hover: text-decoration: underline;">1. Importing the data (Defects)</a></li>
        <li style="margin-bottom: 10px;"><a href="#2" style="text-decoration: none; color: #2a85d0; transition: color 0.3s ease; hover: text-decoration: underline;">2. Defining the class Biscuit and Dough</a>
            <ul style="list-style-type: circle; padding-left: 20px; margin-top: 5px;">
                <li style="margin-bottom: 5px;"><a href="#5" style="text-decoration: none; color: #555; transition: color 0.3s ease; hover: text-decoration: underline;">Class Biscuit</a></li>
                <li style="margin-bottom: 5px;"><a href="#6" style="text-decoration: none; color: #555; transition: color 0.3s ease; hover: text-decoration: underline;">Class Dough</a></li>
            </ul>
        <li style="margin-bottom: 10px;"><a href="#31" style="text-decoration: none; color: #2a85d0; transition: color 0.3s ease; hover: text-decoration: underline;">2.2 Finding the optimum for comparison</a>
        <li style="margin-bottom: 10px;"><a href="#3" style="text-decoration: none; color: #2a85d0; transition: color 0.3s ease; hover: text-decoration: underline;">3. Finding Heuristics</a>
        <li style="margin-bottom: 10px;"><a href="#4" style="text-decoration: none; color: #2a85d0; transition: color 0.3s ease; hover: text-decoration: underline;">4. Debatting Algorithms</a>
            <ul style="list-style-type: circle; padding-left: 20px; margin-top: 5px;">
                <li style="margin-bottom: 5px;"><a href="#7" style="text-decoration: none; color: #555; transition: color 0.3s ease; hover: text-decoration: underline;">Genetic Algorithms</a></li>
                <ul style="list-style-type: square; padding-left: 20px; margin-top: 5px;">
                        <li style="margin-bottom: 5px;"><a href="#10" style="text-decoration: none; color: #777; transition: color 0.3s ease; hover: text-decoration: underline;">(i) Genetic Algorithm</a></li>
                    <li style="margin-bottom: 5px;"><a href="#11" style="text-decoration: none; color: #777; transition: color 0.3s ease; hover: text-decoration: underline;">(ii) Genetic Algorithm with Elitism Roulette</a></li>
                    <li style="margin-bottom: 5px;"><a href="#12" style="text-decoration: none; color: #777; transition: color 0.3s ease; hover: text-decoration: underline;">(iii) Genetic Algorithm with Tournament Selection and Crossover</a></li>
                    <li style="margin-bottom: 5px;"><a href="#20" style="text-decoration: none; color: #777; transition: color 0.3s ease; hover: text-decoration: underline;">(iii) Genetic Algorithm Uniform Crossover</a></li>
                </ul>
                <li style="margin-bottom: 5px;"><a href="#8" style="text-decoration: none; color: #555; transition: color 0.3s ease; hover: text-decoration: underline;">Greedy Searches</a></li>
                <ul style="list-style-type: square; padding-left: 20px; margin-top: 5px;">
                    <li style="margin-bottom: 5px;"><a href="#13" style="text-decoration: none; color: #777; transition: color 0.3s ease; hover: text-decoration: underline;">(i) First heuristic: Value</a></li>
                    <li style="margin-bottom: 5px;"><a href="#14" style="text-decoration: none; color: #777; transition: color 0.3s ease; hover: text-decoration: underline;">(ii) Second  heuristic: Value / Length</a></li>
                    <li style="margin-bottom: 5px;"><a href="#15" style="text-decoration: none; color: #777; transition: color 0.3s ease; hover: text-decoration: underline;">(iii) Third heuristic: Value / (Length*Max number of defects)</a></li>
                    <li style="margin-bottom: 5px;"><a href="#16" style="text-decoration: none; color: #777; transition: color 0.3s ease; hover: text-decoration: underline;">(iv) Fourth heuristic: Without biscuit 2</a></li>
                </ul>
                <li style="margin-bottom: 5px;"><a href="#9" style="text-decoration: none; color: #555; transition: color 0.3s ease; hover: text-decoration: underline;">CSP Algorithms</a></li>
            </ul>
        <li style="margin-bottom: 10px;"><a href="#30" style="text-decoration: none; color: #2a85d0; transition: color 0.3s ease; hover: text-decoration: underline;">Conclusion </a>
    </ul>
</div>


<a id="1"></a>
    
<div class='alert-block alert-info' style="font-size: 24px; font-family: 'Georgia', serif; padding: 2px; border-radius: 10px; margin: 20px; text-align:center; padding-bottom: 20px;">
    <br>
    <h1 align="center"; >1. Importing the data (Defects)</h1>
    <br>
</div>

In [1]:
import pandas as pd
import random
import time

In [2]:
defects_df = pd.read_csv('defects.csv')

<a id="2"></a>
    
<div class='alert-block alert-info' style="font-size: 24px; font-family: 'Georgia', serif; padding: 2px; border-radius: 10px; margin: 20px; text-align:center; padding-bottom: 20px;">
    <br>
    <h1 align="center"; >2.1 Defining the class Biscuit and Dough</h1>
    <br>
</div>

<div style="background-color: #e8f4f8; border: 2px solid #b6d4e0; padding: 20px; border-radius: 10px; margin: 20px; box-shadow: 0 4px 6px rgba(0,0,0,0.1);">
    <h3 style="color: #2c7be5; font-size: 22px; font-family: 'Verdana', sans-serif; margin-bottom: 15px;">Observation  :</h3>
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;">The Biscuit Class is a simple class that only helps vizualizing the particularities of a biscuit: type, length, value, number of defects.
</p>
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;">The Dough Class, is more complex: it helps placing the biscuits in the dough and vizualizing the results (it is explained in the deliverable). It has functions pretty self-explainable (add_defects, count_defects, place_biscuits, GA == Genetic Algorithm, calculate own value, is overlapping, etc.). It is used to know if the biscuits placed are valid and if no overlapping happens too. We use class to simplify our commands and clarity of the code.
</p>
</div>

<a id="5"></a>
<h3 style="color: #4a86e8; font-size: 28px; font-family: 'Georgia', serif; text-shadow: 2px 2px 4px rgba(0,0,0,0.3); margin-top: 20px; border-left: 5px solid #4a86e8; padding-left: 10px;">Class Biscuit</h3>

In [3]:
class Biscuit:
    def __init__(self, biscuit_type):
        self.biscuit_type = biscuit_type
        self.length, self.value, self.max_defects = self._configure_biscuit(biscuit_type)

    def _configure_biscuit(self, biscuit_type):
        if biscuit_type == 0:
            return 4, 6, {'a': 4, 'b': 2, 'c': 3}
        elif biscuit_type == 1:
            return 8, 12, {'a': 5, 'b': 4, 'c': 4}
        elif biscuit_type == 2:
            return 2, 1, {'a': 1, 'b': 2, 'c': 1}
        elif biscuit_type == 3:
            return 5, 8, {'a': 2, 'b': 3, 'c': 2}
        elif biscuit_type == 4:
            return 1, 0, {'a': 10, 'b': 10, 'c': 10} #This biscuit serves for some heuristics
        else:
            return 1, 0, {'a': 10, 'b': 10, 'c': 10}

    def biscuit_type(self):
        return self.biscuit_type
    
    def value(self):
        return self.value
    
    def length(self):
        return self.length
    
    def max_defects(self):
        return self.max_defects
    
    def __str__(self):
        return f"Biscuit Type: {self.biscuit_type}, Length: {self.length}, Value: {self.value}, Max Defects: {self.max_defects}"

In [4]:
biscuit0 = Biscuit(0)
biscuit1 = Biscuit(1)
biscuit2 = Biscuit(2)
biscuit3 = Biscuit(3)
biscuit4 = Biscuit(4)

print(biscuit0)
print(biscuit1)
print(biscuit2)
print(biscuit3)
print(biscuit4)

Biscuit Type: 0, Length: 4, Value: 6, Max Defects: {'a': 4, 'b': 2, 'c': 3}
Biscuit Type: 1, Length: 8, Value: 12, Max Defects: {'a': 5, 'b': 4, 'c': 4}
Biscuit Type: 2, Length: 2, Value: 1, Max Defects: {'a': 1, 'b': 2, 'c': 1}
Biscuit Type: 3, Length: 5, Value: 8, Max Defects: {'a': 2, 'b': 3, 'c': 2}
Biscuit Type: 4, Length: 1, Value: 0, Max Defects: {'a': 10, 'b': 10, 'c': 10}


<a id="6"></a>
<h3 style="color: #4a86e8; font-size: 28px; font-family: 'Georgia', serif; text-shadow: 2px 2px 4px rgba(0,0,0,0.3); margin-top: 20px; border-left: 5px solid #4a86e8; padding-left: 10px;">Class Dough</h3>

In [5]:
class Dough:
    def __init__(self, length):
        self.LENGTH = length
        self.defects_list = []  # tuples (position, class)
        self.biscuits_placed = []  # tuples (position, biscuit_type, valid)

    def add_defect(self, position, defect_class):
        if 0 <= position < self.LENGTH:
            self.defects_list.append((position, defect_class))
        else:
            raise ValueError("Defect position out of dough range")

    def count_defects(self, position, length):
        defect_counts = {}
        for pos, cls in self.defects_list:
            if position <= pos < position + length:
                defect_counts[cls] = defect_counts.get(cls, 0) + 1
        return defect_counts

    def place_biscuits(self, biscuits):
        self.biscuits_placed.clear()
        current_position = 0
        for biscuit in biscuits:
            if current_position + biscuit.length > self.LENGTH:
                break  # Stop if there's no enough space left for the current biscuit
    
            defect_counts = self.count_defects(current_position, biscuit.length)
            valid = all(defect_counts.get(cls, 0) <= biscuit.max_defects[cls] for cls in biscuit.max_defects)
            self.biscuits_placed.append((current_position, biscuit.biscuit_type, valid))
            current_position += biscuit.length
            self.biscuits_placed = sorted(self.biscuits_placed, key=lambda x: x[0])


    def place_biscuits_GA(self, biscuit_positions):
        self.biscuits_placed.clear()
        for position, biscuit_type in biscuit_positions:
            biscuit = Biscuit(biscuit_type)
            if position + biscuit.length > self.LENGTH:
                break  
            defect_counts = self.count_defects(position, biscuit.length)
            valid = all(defect_counts.get(cls, 0) <= biscuit.max_defects[cls] for cls in biscuit.max_defects)
            self.biscuits_placed.append((position, biscuit.biscuit_type, valid))
            self.biscuits_placed = sorted(self.biscuits_placed, key=lambda x: x[0])

    
    def get_positions(self):
        for position, biscuit_type, is_valid in self.biscuits_placed:
            print(f"biscuit type: {biscuit_type}, position: {position}")
    
    def calculate_total_value(self, solution):
        total_value = 0
        for position, biscuit_type, is_valid in solution:
            defect = self.count_defects(current_position, biscuit.length)
            print(f"position: {position}, biscuit_type: {biscuit_type}, defect: {defect}, isValid: {is_valid}")
            if is_valid:
                total_value += self.biscuits[biscuit_type].value
        return total_value

    def calculate_own_value(self, biscuits):
        total_value = 0
        for position, biscuit_type, is_valid in self.biscuits_placed:
            if is_valid:
                total_value += biscuits[biscuit_type].value
        return total_value

    def validate_biscuits_placement(self, biscuits):
        last_position = -1
        for position, biscuit_type, is_valid in self.biscuits_placed:
            biscuit = biscuits[biscuit_type]

            # Checks for overlapping
            if position < last_position:
                return False, f"Overlapping detected at position {position} for biscuit type {biscuit_type}."
            last_position = position + biscuit.length

        return True, "All biscuits are correctly placed and within defect thresholds."

    def is_overlapping(self, new_position, new_length, existing_biscuits):
        for position, biscuit_type in existing_biscuits:
            biscuit_length = self.biscuits[biscuit_type].length
            if not (new_position + new_length <= position or new_position >= position + biscuit_length):
                return True 
        return False 

    def __str__(self):
        return ("Dough Defects:\n" + "\n".join(f"Position: {pos:.2f}, Class: {cls}" for pos, cls in self.defects_list) +
                "\n\nBiscuits Placed:\n" + "\n".join(f"Position: {pos}, Biscuit Type: {b_type}, Valid: {valid}" 
                                                      for pos, b_type, valid in self.biscuits_placed))

    def defects(self):
        return self.defects_list

    def biscuits(self):
        return self.biscuits_placed

In [6]:
dough = Dough(length=500)

for _, row in defects_df.iterrows():
    dough.add_defect(position=row['x'], defect_class=row['class'])

In [7]:
print(dough)

Dough Defects:
Position: 355.45, Class: c
Position: 92.50, Class: a
Position: 141.88, Class: c
Position: 431.83, Class: c
Position: 435.03, Class: c
Position: 205.80, Class: a
Position: 34.69, Class: b
Position: 443.57, Class: a
Position: 69.42, Class: a
Position: 301.28, Class: a
Position: 342.65, Class: c
Position: 216.12, Class: a
Position: 367.67, Class: c
Position: 93.13, Class: a
Position: 401.42, Class: c
Position: 74.25, Class: c
Position: 356.84, Class: a
Position: 457.42, Class: a
Position: 395.64, Class: c
Position: 368.89, Class: a
Position: 163.92, Class: a
Position: 484.16, Class: b
Position: 55.27, Class: c
Position: 171.56, Class: b
Position: 119.69, Class: a
Position: 332.64, Class: b
Position: 85.10, Class: a
Position: 289.68, Class: b
Position: 156.79, Class: b
Position: 143.53, Class: c
Position: 83.61, Class: a
Position: 108.52, Class: a
Position: 57.58, Class: c
Position: 9.96, Class: c
Position: 440.00, Class: b
Position: 146.01, Class: b
Position: 259.33, Class:

<a id="31"></a>
    
<div class='alert-block alert-info' style="font-size: 24px; font-family: 'Georgia', serif; padding: 2px; border-radius: 10px; margin: 20px; text-align:center; padding-bottom: 20px;">
    <br>
    <h1 align="center"; >2.2 Finding the optimum for comparison</h1>
    <br>
</div>

<div style="background-color: #e8f4f8; border: 2px solid #b6d4e0; padding: 20px; border-radius: 10px; margin: 20px; box-shadow: 0 4px 6px rgba(0,0,0,0.1);">
    <h3 style="color: #2c7be5; font-size: 22px; font-family: 'Verdana', sans-serif; margin-bottom: 15px;">Dynamic programming approach  :</h3>
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;">This problem can be solved using a dynamic programming approach in <strong>O(n) time complexity and O(n^2) space complexity</strong>. In fact we know that the goal is to use algorithms that we saw in class but as a reference it is really usefull to find the value of the best appoach as a comparison.
</p>
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;"><strong>Here is the main idea :</strong>
</p>
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;"><strong>The best possible value at the i-th position, \( V(i) \), is the maximum value achievable by considering all possible placements of biscuits up to the i-th position. This can be calculated as the maximum value obtained by placing a biscuit of length \( L \) at the current position, and adding it to the best value at the position \( i - L \). 

where \( L \) varies over all possible biscuit lengths.
</strong>
</p>
</div></div>

In [8]:
#Variables
baguette  = [[] for i in range(dough.LENGTH)] #Use to simplify notation here, represents defect on their positions
for _, row in defects_df.iterrows():
    baguette[int(row['x'])] = baguette[int(row['x'])]+[row['class']]
    
biscuit_types = [Biscuit(i) for i in range(4)]# all 4 biscuits type used here

maxAtindex = [[] for i in range(500)] #use to keep track of the best path at ith position

In [9]:
#Usefull functions :
def get_val(roll):#get the val from a given agencement of biscuits (sum their values)
    s=0
    for b,pos in roll:
        s+=b.value
    return s

def canBePlaced(biscuit,i,baguette = baguette): #return True if biscuit can be placed at index i on baguette 
    ca = 0
    cb = 0
    cc = 0
    for index in range(i,biscuit.length+i):
        ca+=baguette[index].count('a')
        cb+=baguette[index].count('b')
        cc+=baguette[index].count('c')
    if  ca<= biscuit.max_defects['a'] and  cb<= biscuit.max_defects['b'] and  cc<= biscuit.max_defects['c']:
        return True
    else:
        return False

In [10]:
start = time.time()
for i in range(dough.LENGTH):
    maxVal = float("-inf")
    maxroll = []
    for b in biscuit_types:#iterate over all biscuits types
        if i>=b.length:
            roll = maxAtindex[i-b.length]#current best roll at position i-L
            currentVal = get_val(roll)
            
            if canBePlaced(b,i-b.length):#check if we can place it
                #print(b,i-b.length)
                currentVal+=b.value
            #print(currentVal)
            if currentVal>=maxVal:#if we got a better current value we replace it
                
                maxroll = roll+[(b,i-b.length)]
                maxVal = currentVal
    #print(maxroll)
    maxAtindex[i] = maxroll
print("Found in : ",time.time()-start,"seconds")
print("Value of Maximum : " , get_val(maxAtindex[-1]))
for b,x in maxAtindex[-1] : 
    print(b, " at position : ",x)

Found in :  0.028316259384155273 seconds
Value of Maximum :  760
Biscuit Type: 0, Length: 4, Value: 6, Max Defects: {'a': 4, 'b': 2, 'c': 3}  at position :  1
Biscuit Type: 1, Length: 8, Value: 12, Max Defects: {'a': 5, 'b': 4, 'c': 4}  at position :  5
Biscuit Type: 3, Length: 5, Value: 8, Max Defects: {'a': 2, 'b': 3, 'c': 2}  at position :  13
Biscuit Type: 1, Length: 8, Value: 12, Max Defects: {'a': 5, 'b': 4, 'c': 4}  at position :  18
Biscuit Type: 3, Length: 5, Value: 8, Max Defects: {'a': 2, 'b': 3, 'c': 2}  at position :  26
Biscuit Type: 3, Length: 5, Value: 8, Max Defects: {'a': 2, 'b': 3, 'c': 2}  at position :  31
Biscuit Type: 3, Length: 5, Value: 8, Max Defects: {'a': 2, 'b': 3, 'c': 2}  at position :  36
Biscuit Type: 0, Length: 4, Value: 6, Max Defects: {'a': 4, 'b': 2, 'c': 3}  at position :  41
Biscuit Type: 3, Length: 5, Value: 8, Max Defects: {'a': 2, 'b': 3, 'c': 2}  at position :  45
Biscuit Type: 0, Length: 4, Value: 6, Max Defects: {'a': 4, 'b': 2, 'c': 3}  at 

<div style="background-color: #e8f4f8; border: 2px solid #b6d4e0; padding: 20px; border-radius: 10px; margin: 20px; box-shadow: 0 4px 6px rgba(0,0,0,0.1);">
    <h3 style="color: #2c7be5; font-size: 22px; font-family: 'Verdana', sans-serif; margin-bottom: 15px;">Observation  :</h3>
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;">We can observe that using this approach we find an optimum in 0.014 seconds which is really fast.(Note that we could have listed all optimums by keeping track of egality cases). Now we know that the <strong>optimum value is 760</strong> so let's try to beat approach it as close as possible with others methods !
</p>

</div>

<a id="3"></a>
    
<div class='alert-block alert-info' style="font-size: 24px; font-family: 'Georgia', serif; padding: 2px; border-radius: 10px; margin: 20px; text-align:center; padding-bottom: 20px;">
    <br>
    <h1 align="center"; >3. Finding Heuristics</h1>
    <br>
</div>

<body style="font-family: Arial, sans-serif; line-height: 1.6;">
    <div style="margin: 0 auto; width: 80%; max-width: 700px; padding: 20px;">
        <h2 style="border-bottom: 2px solid #333; padding-bottom: 10px;">Objective Function</h2>
        <p>We want to:</p>
        <ul style="list-style-type: none; padding: 0;">
            <li style="margin-bottom: 10px; display: list-item; list-style-position: inside;">Maximize the value (=<span> Sum of the biscuits' values on the dough</span>)</li>
        </ul>
        <p>Maximizing the value means:</p>
        <ul style="list-style-type: none; padding: 0;">
            <li style="margin-bottom: 10px; display: list-item; list-style-position: inside;">Minimizing the value of the lost biscuits (biscuits with more defects than threshold)</li>
            <li style="display: list-item; list-style-position: inside;">Maximize the biscuits with the most values</li>
        </ul>
    </div>
</body>

In [8]:
biscuits = [biscuit0,biscuit1,biscuit2,biscuit3]

In [9]:
for biscuit in biscuits:
    print("Biscuit type: ",biscuit.biscuit_type)
    print("Value/Length: ",biscuit.value/biscuit.length)
    maxd = sum(biscuit.max_defects.values())
    print("Max defects: ", maxd)
    print("Value/(Length*Max_Defects): ", biscuit.value/(biscuit.length*maxd), "\n")

Biscuit type:  0
Value/Length:  1.5
Max defects:  9
Value/(Length*Max_Defects):  0.16666666666666666 

Biscuit type:  1
Value/Length:  1.5
Max defects:  13
Value/(Length*Max_Defects):  0.11538461538461539 

Biscuit type:  2
Value/Length:  0.5
Max defects:  4
Value/(Length*Max_Defects):  0.125 

Biscuit type:  3
Value/Length:  1.6
Max defects:  7
Value/(Length*Max_Defects):  0.22857142857142856 



Our heuristic tells us the order:
>Biscuit 3 > Biscuit 0 > Biscuit 1 > Biscuit 2

Therefore, we want to:
- Have the max Biscuits placed
- Respect our order priority so that the value is maxed out

<div style="background-color: #e8f4f8; border: 2px solid #b6d4e0; padding: 20px; border-radius: 10px; margin: 20px; box-shadow: 0 4px 6px rgba(0,0,0,0.1);">
    <h3 style="color: #2c7be5; font-size: 22px; font-family: 'Verdana', sans-serif; margin-bottom: 15px;">Observation  :</h3>
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;">Although this is an heuristic, we might find ourselves using another heuristic if we conclude that it works better. In a Problem with static contraints like the defects here, sometimes the most rational heuristic is not the most optimal. Therefore it is our duty to find the best one.
</p>

</div>

<a id="4"></a>
    
<div class='alert-block alert-info' style="font-size: 24px; font-family: 'Georgia', serif; padding: 2px; border-radius: 10px; margin: 20px; text-align:center; padding-bottom: 20px;">
    <br>
    <h1 align="center"; >4. Debatting Algorithms</h1>
    <br>
</div>

<a id="7"></a>
<h3 style="color: #4a86e8; font-size: 28px; font-family: 'Georgia', serif; text-shadow: 2px 2px 4px rgba(0,0,0,0.3); margin-top: 20px; border-left: 5px solid #4a86e8; padding-left: 10px;">Genetic Algorithm</h3>

<a id="10"><h4 style="color: #6a737b; font-size: 24px; font-family: 'Courier New', monospace; margin-top: 15px; border-bottom: 1px dashed #6a737b; padding-bottom: 5px;">(i) Class Genetic Algorithm</h4></a>

<div style="background-color: #e8f4f8; border: 2px solid #b6d4e0; padding: 20px; border-radius: 10px; margin: 20px; box-shadow: 0 4px 6px rgba(0,0,0,0.1);">
    <h3 style="color: #2c7be5; font-size: 22px; font-family: 'Verdana', sans-serif; margin-bottom: 15px;">Remarks:</h3>
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;">This is the basis of our GA reasoning, which means we will implement sub-classes from this one as a upper class. Our results will vary from the original functions emanating from this class. You can find its description in the deliverable, but there are comments resuming each function.
</p>

</div>

In [109]:
class GeneticAlgorithm:

    #Obviously initialisation
    def __init__(self, dough, biscuits, population_size, mutation_rate, crossover_rate):
        self.dough = dough
        self.biscuits = biscuits
        self.population_size = population_size
        self.mutation_rate = mutation_rate
        self.crossover_rate = crossover_rate
        self.population = []
        self.initialize_population()

    #works with random solution to generate populations totally randomly
    def initialize_population(self):
        for _ in range(self.population_size):
            individual = self.random_solution()
            self.population.append(individual)

    #Generates a random solution, with more chances to summon biscuits 3 than 0s and 1s than 2, as our heuristic tells us
    def random_solution(self):
        biscuits = [3, 3, 3, 3] + [0, 0, 1, 1] + [2]
        solution = []
        position = 0
        while position < self.dough.LENGTH:
            biscuit_type = random.choice(biscuits)
            solution.append((position, biscuit_type))
            position += self.biscuits[biscuit_type].length
        return solution

    #Calculates the value of an individual, we did not go through the class Dough because that would make us lose Space Complexity.
    def fitness(self, individual):
        total_value = 0
        for position, biscuit_type in individual:
            biscuit = self.biscuits[biscuit_type]
            # checks if biscuit fits in the remaining dough and meets defect requirements
            if position + biscuit.length <= self.dough.LENGTH:
                defects = self.dough.count_defects(position, biscuit.length)
                if all(defects.get(cls, 0) <= biscuit.max_defects[cls] for cls in biscuit.max_defects):
                    total_value += biscuit.value
        return total_value

    #simple roulette wheel selection
    def selection(self):
        max_fitness = sum(self.fitness(ind) for ind in self.population)
        pick = random.uniform(0, max_fitness)
        current = 0
        for individual in self.population:
            current += self.fitness(individual)
            if current > pick:
                return individual

    #This is a ONE POINT Crossover, that makes sure that none of the "cuts" made to combine the parents cuts a biscuit in half and causes overlaps
    @staticmethod
    def crossover(parent1, parent2, max_attempts = 500):
        parent1 = sorted(parent1, key=lambda x: x[0])
        parent2 = sorted(parent2, key=lambda x: x[0])
        biscuits = {
            0: Biscuit(0),
            1: Biscuit(1),
            2: Biscuit(2),
            3: Biscuit(3),
            4: Biscuit(4)
        }
        #Finds the valid points
        def find_valid_crossover_points(parent, biscuits):
            valid_points = []
            for i in range(1, len(parent)):
                prev_position, prev_type = parent[i-1]
                position, _ = parent[i]
                if position - (prev_position + biscuits[prev_type].length) > 0:
                    valid_points.append(i)
            return valid_points

        #Creates the child with the found crossover point
        def create_child(p1, p2, crossover_point):
            child = [None] * len(p1)
            child[:crossover_point] = p1[:crossover_point]
            child[crossover_point:] = [b for b in p2[crossover_point:] if b not in child[:crossover_point]]
            return child

        for _ in range(max_attempts):
            valid_points_parent1 = find_valid_crossover_points(parent1, biscuits)
            valid_points_parent2 = find_valid_crossover_points(parent2, biscuits)

            if valid_points_parent1 and valid_points_parent2:
                crossover_point1 = random.choice(valid_points_parent1)
                crossover_point2 = random.choice(valid_points_parent2)

                child1 = create_child(parent1, parent2, crossover_point1)
                child2 = create_child(parent2, parent1, crossover_point2)

                return child1, child2
        #In case no valid point is found:
        return parent1, parent2

    #To keep diversity, we need to insert mutation:
    # simple mutation where a random biscuit's type or position might change
    def mutate(self, individual):
        if len(individual) > 0 and random.random() < self.mutation_rate:
            mutate_index = random.randint(0, len(individual) - 1)
            if random.random() > 0.5:
                individual[mutate_index] = (individual[mutate_index][0], random.choice(list(self.biscuits.keys())))
            else:
                position = individual[mutate_index][0]
                new_position = max(0, min(self.dough.LENGTH - self.biscuits[individual[mutate_index][1]].length, position + random.choice([-1, 1])))
                individual[mutate_index] = (new_position, individual[mutate_index][1])
        return individual
        
    #Defines how the population will evolve through generations
    def evolve(self):
        new_population = []
        while len(new_population) < self.population_size:
            parent1 = self.selection()
            parent2 = self.selection()
            if random.random() < self.crossover_rate:
                offspring1, offspring2 = self.crossover(parent1, parent2)
            else:
                offspring1, offspring2 = parent1, parent2
            offspring1 = self.mutate(offspring1)
            offspring2 = self.mutate(offspring2)
            new_population.extend([offspring1, offspring2])
        self.population = new_population[:self.population_size] 

In [12]:
#Simple program to launch the Genetic Algorithm
def main():
    dough = Dough(length=500)
    for _, row in defects_df.iterrows():
        dough.add_defect(position=row['x'], defect_class=row['class'])
        
    biscuits = {
        0: Biscuit(0),
        1: Biscuit(1),
        2: Biscuit(2),
        3: Biscuit(3)
    }
    
    GA = GeneticAlgorithm(dough, biscuits, population_size=50, mutation_rate=0.01, crossover_rate=0.7)
    GA.initialize_population()

    for generation in range(100): 
        #print(generation) # Run for 100 generations or until termination criteria are met
        GA.evolve()

    best_solution = max(GA.population, key=GA.fitness)
    total_value = GA.fitness(best_solution)
    # [biscuit1, 0, 0, biscuit2]
    #dough.calculate_total_value(best_solution)
    print("Total Value:", total_value)
    print("Best Solution:", best_solution)
    
    return best_solution

if __name__ == "__main__":
    best_solution = main()

Total Value: 523
Best Solution: [(0, 1), (8, 3), (13, 3), (18, 3), (23, 2), (25, 3), (30, 1), (38, 3), (43, 2), (45, 3), (50, 1), (58, 1), (66, 0), (70, 2), (72, 3), (77, 3), (82, 0), (86, 3), (91, 0), (95, 1), (103, 1), (111, 1), (119, 3), (124, 3), (129, 2), (131, 3), (136, 1), (144, 0), (148, 0), (152, 1), (160, 3), (165, 2), (167, 1), (175, 0), (179, 3), (184, 3), (189, 3), (194, 2), (196, 3), (201, 0), (205, 1), (213, 0), (217, 1), (225, 3), (230, 0), (234, 2), (236, 2), (238, 0), (242, 3), (247, 1), (255, 3), (260, 1), (268, 3), (273, 3), (278, 3), (283, 2), (285, 0), (289, 1), (297, 2), (299, 0), (303, 0), (307, 1), (315, 3), (320, 3), (325, 0), (329, 1), (337, 1), (345, 1), (353, 0), (357, 1), (365, 2), (367, 3), (372, 3), (377, 3), (382, 3), (387, 2), (389, 0), (391, 1), (393, 1), (399, 2), (401, 0), (401, 1), (405, 1), (409, 2), (411, 1), (413, 0), (417, 0), (419, 3), (421, 0), (424, 3), (429, 0), (433, 0), (437, 1), (445, 3), (450, 1), (458, 0), (462, 2), (464, 3), (469, 3),

<div style="background-color: #e8f4f8; border: 2px solid #b6d4e0; padding: 20px; border-radius: 10px; margin: 20px; box-shadow: 0 4px 6px rgba(0,0,0,0.1);">
    <h3 style="color: #2c7be5; font-size: 22px; font-family: 'Verdana', sans-serif; margin-bottom: 15px;">Observation  :</h3>
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;">The output value is preetty low, but that could be expected since our population is pretty small (50) and the number of defects far exceeds that, we would need a bigger population to reinforce our results or increase our mutation rates to find by hazard a better solution
</p>

</div>

In [13]:
dough = Dough(length=500)
for _, row in defects_df.iterrows():
    dough.add_defect(position=row['x'], defect_class=row['class'])
    
dough.place_biscuits_GA(best_solution)
print(dough)

Dough Defects:
Position: 355.45, Class: c
Position: 92.50, Class: a
Position: 141.88, Class: c
Position: 431.83, Class: c
Position: 435.03, Class: c
Position: 205.80, Class: a
Position: 34.69, Class: b
Position: 443.57, Class: a
Position: 69.42, Class: a
Position: 301.28, Class: a
Position: 342.65, Class: c
Position: 216.12, Class: a
Position: 367.67, Class: c
Position: 93.13, Class: a
Position: 401.42, Class: c
Position: 74.25, Class: c
Position: 356.84, Class: a
Position: 457.42, Class: a
Position: 395.64, Class: c
Position: 368.89, Class: a
Position: 163.92, Class: a
Position: 484.16, Class: b
Position: 55.27, Class: c
Position: 171.56, Class: b
Position: 119.69, Class: a
Position: 332.64, Class: b
Position: 85.10, Class: a
Position: 289.68, Class: b
Position: 156.79, Class: b
Position: 143.53, Class: c
Position: 83.61, Class: a
Position: 108.52, Class: a
Position: 57.58, Class: c
Position: 9.96, Class: c
Position: 440.00, Class: b
Position: 146.01, Class: b
Position: 259.33, Class:

In [15]:
biscuits_info = {
        0: Biscuit(0),
        1: Biscuit(1),
        2: Biscuit(2),
        3: Biscuit(3),
        4: Biscuit(4)
    }

In [16]:
dough.calculate_own_value(biscuits_info)

523

In [32]:
dough.validate_biscuits_placement(biscuits_info)

(True, 'All biscuits are correctly placed and within defect thresholds.')

<a id="11"><h4 style="color: #6a737b; font-size: 24px; font-family: 'Courier New', monospace; margin-top: 15px; border-bottom: 1px dashed #6a737b; padding-bottom: 5px;">(ii) Genetic Algorithm with Elitism Roulette</h4></a>

<div style="background-color: #e8f4f8; border: 2px solid #b6d4e0; padding: 20px; border-radius: 10px; margin: 20px; box-shadow: 0 4px 6px rgba(0,0,0,0.1);">
    <h3 style="color: #2c7be5; font-size: 22px; font-family: 'Verdana', sans-serif; margin-bottom: 15px;">Remarks  :</h3>
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;">We create a sub-class to Genetic Algorithm that uses elitism to find the next generation, meaning it uses the best individuals (20% of the population) to perform the crossovers and surely ensure a better solution.
</p>

</div>

In [20]:
class GeneticElitism(GeneticAlgorithm):
    def __init__(self, dough, biscuits, population_size, mutation_rate, crossover_rate):
        super().__init__(dough, biscuits, population_size, mutation_rate, crossover_rate)

    def selection(self):
        max_fitness = sum(self.fitness(ind) for ind in self.population)
        pick = random.uniform(0, max_fitness)
        current = 0
        for individual in self.population:
            current += self.fitness(individual)
            if current > pick:
                return individual

In [22]:
def main():
    dough = Dough(length=500) 
    for _, row in defects_df.iterrows():
        dough.add_defect(position=row['x'], defect_class=row['class'])
    biscuits = {
        0: Biscuit(0),
        1: Biscuit(1),
        2: Biscuit(2),
        3: Biscuit(3)
    }
    
    GA = GeneticElitism(dough, biscuits, population_size=50, mutation_rate=0.01, crossover_rate=0.7)
    GA.initialize_population()

    for generation in range(100): 
        #print(generation) # Run for 100 generations or until termination criteria are met
        GA.evolve()

    # Find the best solution in the final population
    best_solution = max(GA.population, key=GA.fitness)
    total_value = GA.fitness(best_solution)
    # [biscuit1, 0, 0, biscuit2]
    #dough.calculate_total_value(best_solution)
    print("Total Value:", total_value)
    print("Best Solution:", best_solution)
    return best_solution

if __name__ == "__main__":
    best_solution = main()

Total Value: 537
Best Solution: [(0, 1), (8, 1), (16, 0), (20, 1), (28, 3), (33, 2), (35, 1), (43, 3), (48, 3), (53, 1), (61, 2), (63, 2), (65, 0), (69, 1), (77, 0), (81, 3), (86, 3), (91, 2), (93, 3), (98, 1), (106, 0), (110, 0), (114, 1), (122, 2), (124, 0), (128, 3), (133, 3), (138, 2), (140, 2), (142, 2), (144, 1), (152, 2), (154, 3), (159, 1), (167, 1), (175, 1), (183, 3), (188, 3), (193, 2), (195, 0), (199, 1), (207, 0), (211, 0), (215, 0), (219, 1), (227, 1), (235, 3), (240, 2), (242, 0), (246, 3), (251, 2), (253, 2), (255, 3), (260, 1), (268, 0), (272, 0), (276, 1), (284, 0), (288, 3), (293, 1), (301, 3), (306, 0), (310, 1), (318, 3), (323, 3), (328, 3), (333, 1), (341, 0), (345, 3), (350, 3), (355, 0), (359, 0), (363, 0), (367, 2), (369, 1), (377, 2), (379, 1), (387, 0), (391, 0), (395, 2), (397, 3), (402, 3), (407, 1), (415, 3), (420, 1), (428, 1), (436, 2), (438, 2), (440, 3), (445, 2), (447, 3), (449, 2), (451, 1), (459, 0), (463, 3), (468, 0), (472, 3), (477, 3), (482, 3),

In [23]:
#Juste un test pour placer les biscuits sur le dough avec le GA et vérifier que tout fonctionnait
dough = Dough(length=500)
for _, row in defects_df.iterrows():
    dough.add_defect(position=row['x'], defect_class=row['class'])
dough.place_biscuits_GA(best_solution)
print(dough)

Dough Defects:
Position: 355.45, Class: c
Position: 92.50, Class: a
Position: 141.88, Class: c
Position: 431.83, Class: c
Position: 435.03, Class: c
Position: 205.80, Class: a
Position: 34.69, Class: b
Position: 443.57, Class: a
Position: 69.42, Class: a
Position: 301.28, Class: a
Position: 342.65, Class: c
Position: 216.12, Class: a
Position: 367.67, Class: c
Position: 93.13, Class: a
Position: 401.42, Class: c
Position: 74.25, Class: c
Position: 356.84, Class: a
Position: 457.42, Class: a
Position: 395.64, Class: c
Position: 368.89, Class: a
Position: 163.92, Class: a
Position: 484.16, Class: b
Position: 55.27, Class: c
Position: 171.56, Class: b
Position: 119.69, Class: a
Position: 332.64, Class: b
Position: 85.10, Class: a
Position: 289.68, Class: b
Position: 156.79, Class: b
Position: 143.53, Class: c
Position: 83.61, Class: a
Position: 108.52, Class: a
Position: 57.58, Class: c
Position: 9.96, Class: c
Position: 440.00, Class: b
Position: 146.01, Class: b
Position: 259.33, Class:

In [24]:
biscuits_info = {
        0: Biscuit(0),
        1: Biscuit(1),
        2: Biscuit(2),
        3: Biscuit(3)
    }
dough.calculate_own_value(biscuits_info)

537

<div style="background-color: #e8f4f8; border: 2px solid #b6d4e0; padding: 20px; border-radius: 10px; margin: 20px; box-shadow: 0 4px 6px rgba(0,0,0,0.1);">
    <h3 style="color: #2c7be5; font-size: 22px; font-family: 'Verdana', sans-serif; margin-bottom: 15px;">Observation  :</h3>
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;">Our results stabilize at a higher value than before but not much. We can think that is logic since that the population still have the same variety, we only increased the speed of the algorithm.
</p>

</div>

In [33]:
dough.validate_biscuits_placement(biscuits_info)

(True, 'All biscuits are correctly placed and within defect thresholds.')

<a id="12"><h4 style="color: #6a737b; font-size: 24px; font-family: 'Courier New', monospace; margin-top: 15px; border-bottom: 1px dashed #6a737b; padding-bottom: 5px;">(iii) Genetic Algorithm with Tournament Selection and Crossover</h4></a>

<div style="background-color: #e8f4f8; border: 2px solid #b6d4e0; padding: 20px; border-radius: 10px; margin: 20px; box-shadow: 0 4px 6px rgba(0,0,0,0.1);">
    <h3 style="color: #2c7be5; font-size: 22px; font-family: 'Verdana', sans-serif; margin-bottom: 15px;">Notice  :</h3>
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;">Using tournament selection system, we take samples of the population ranked to sort out the results
</p>

</div>

In [110]:
class GeneticTournament(GeneticAlgorithm):
    def __init__(self, dough, biscuits, population_size, mutation_rate, crossover_rate):
        super().__init__(dough, biscuits, population_size, mutation_rate, crossover_rate)
        self.tournament_size = 5
        
    def selection(self, elite_size):
        pop_ranked = [(individual, self.fitness(individual)) for individual in self.population]
        pop_ranked.sort(key=lambda x: x[1], reverse=True)
        selection_results = [individual[0] for individual in pop_ranked[:elite_size]]
        tournament_size = 10
        
        for _ in range(len(pop_ranked) - elite_size):
            tournament = random.sample(pop_ranked, tournament_size)
            selection_results.append(sorted(tournament, key=lambda x: x[1], reverse=True)[0][0])
        return selection_results
        
    @staticmethod
    def mutate(individual, mutation_rate):
        for swapped in range(len(individual)):
            if random.random() < mutation_rate:
                swap_with = random.randint(0, len(individual) - 1)
                individual[swapped], individual[swap_with] = individual[swap_with], individual[swapped]
        return individual

    def mutate_population(self, population):
        return [self.mutate(ind, self.mutation_rate) for ind in population]

    def evolve(self):
        elite_size = int(self.population_size * 0.2)
        selected = self.selection(elite_size)
        offspring = []
        for i in range(0, self.population_size - elite_size, 2):
            parent1 = selected[i % len(selected)]
            parent2 = selected[(i+1) % len(selected)]
            child1, child2 = GeneticAlgorithm.crossover(parent1, parent2)
            offspring.append(child1)
            if len(offspring) < self.population_size - elite_size:
                offspring.append(child2)
        self.population = selected[:elite_size] + self.mutate_population(offspring)

In [111]:
def main():
    biscuits = {
        0: Biscuit(0),
        1: Biscuit(1),
        2: Biscuit(2),
        3: Biscuit(3),
        4: Biscuit(4)
    }
    dough = Dough(500)

    for _, row in defects_df.iterrows():
        dough.add_defect(position=row['x'], defect_class=row['class'])

    population_size = 150
    mutation_rate = 0.1
    crossover_rate = 0.5
    GA = GeneticTournament(dough, biscuits, population_size, mutation_rate, crossover_rate)
    num_generations = 100
    
    for generation in range(num_generations):
        GA.evolve()
        print(f"Generation {generation + 1} complete.")

    best_solution = max(GA.population, key=GA.fitness)
    total_value = GA.fitness(best_solution)
    print("Best Solution Total Value:", total_value)
    print("Best Solution Configuration:", best_solution)
    return best_solution

if __name__ == "__main__":
    best_solution = main()

Generation 1 complete.
Generation 2 complete.
Generation 3 complete.
Generation 4 complete.
Generation 5 complete.
Generation 6 complete.
Generation 7 complete.
Generation 8 complete.
Generation 9 complete.
Generation 10 complete.
Generation 11 complete.
Generation 12 complete.
Generation 13 complete.
Generation 14 complete.
Generation 15 complete.
Generation 16 complete.
Generation 17 complete.
Generation 18 complete.
Generation 19 complete.
Generation 20 complete.
Generation 21 complete.
Generation 22 complete.
Generation 23 complete.
Generation 24 complete.
Generation 25 complete.
Generation 26 complete.
Generation 27 complete.
Generation 28 complete.
Generation 29 complete.
Generation 30 complete.
Generation 31 complete.
Generation 32 complete.
Generation 33 complete.
Generation 34 complete.
Generation 35 complete.
Generation 36 complete.
Generation 37 complete.
Generation 38 complete.
Generation 39 complete.
Generation 40 complete.
Generation 41 complete.
Generation 42 complete.
G

In [112]:
biscuits = {
        0: Biscuit(0),
        1: Biscuit(1),
        2: Biscuit(2),
        3: Biscuit(3)
    }
dough = Dough(500)
for _, row in defects_df.iterrows():
    dough.add_defect(position=row['x'], defect_class=row['class'])
dough.place_biscuits_GA(best_solution)
print(dough)

Dough Defects:
Position: 355.45, Class: c
Position: 92.50, Class: a
Position: 141.88, Class: c
Position: 431.83, Class: c
Position: 435.03, Class: c
Position: 205.80, Class: a
Position: 34.69, Class: b
Position: 443.57, Class: a
Position: 69.42, Class: a
Position: 301.28, Class: a
Position: 342.65, Class: c
Position: 216.12, Class: a
Position: 367.67, Class: c
Position: 93.13, Class: a
Position: 401.42, Class: c
Position: 74.25, Class: c
Position: 356.84, Class: a
Position: 457.42, Class: a
Position: 395.64, Class: c
Position: 368.89, Class: a
Position: 163.92, Class: a
Position: 484.16, Class: b
Position: 55.27, Class: c
Position: 171.56, Class: b
Position: 119.69, Class: a
Position: 332.64, Class: b
Position: 85.10, Class: a
Position: 289.68, Class: b
Position: 156.79, Class: b
Position: 143.53, Class: c
Position: 83.61, Class: a
Position: 108.52, Class: a
Position: 57.58, Class: c
Position: 9.96, Class: c
Position: 440.00, Class: b
Position: 146.01, Class: b
Position: 259.33, Class:

In [113]:
biscuits_info = {
        0: Biscuit(0),
        1: Biscuit(1),
        2: Biscuit(2),
        3: Biscuit(3),
        4: Biscuit(4)
    }

In [114]:
dough.calculate_own_value(biscuits_info)

540

In [115]:
dough.validate_biscuits_placement(biscuits_info)

(True, 'All biscuits are correctly placed and within defect thresholds.')

<div style="background-color: #e8f4f8; border: 2px solid #b6d4e0; padding: 20px; border-radius: 10px; margin: 20px; box-shadow: 0 4px 6px rgba(0,0,0,0.1);">
    <h3 style="color: #2c7be5; font-size: 22px; font-family: 'Verdana', sans-serif; margin-bottom: 15px;">Observation  :</h3>
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;">With a bigger population, we equalize the Tournament Elitism selections methods, with a score around 540. Thus, each step of verification is correct. Now, we will try to change our method of crossover:
</p>

</div>

<a id="20"><h4 style="color: #6a737b; font-size: 24px; font-family: 'Courier New', monospace; margin-top: 15px; border-bottom: 1px dashed #6a737b; padding-bottom: 5px;">(iv) Uniform Crossover</h4></a>

In [128]:
class UniformCrossoverGA(GeneticAlgorithm):
    
    def __init__(self, dough, biscuits, population_size, mutation_rate, crossover_rate):
        super().__init__(dough, biscuits, population_size, mutation_rate, crossover_rate)
        self.tournament_size = 15

    def selection(self, elite_size):
        pop_ranked = [(individual, self.fitness(individual)) for individual in self.population]
        pop_ranked.sort(key=lambda x: x[1], reverse=True)
        selection_results = [individual[0] for individual in pop_ranked[:elite_size]]
        tournament_size = 15
        
        for _ in range(len(pop_ranked) - elite_size):
            tournament = random.sample(pop_ranked, tournament_size)
            selection_results.append(sorted(tournament, key=lambda x: x[1], reverse=True)[0][0])
        return selection_results

    @staticmethod
    def uniform_crossover(parent1, parent2, max_attempts=500):
        biscuits = {
            0: Biscuit(0),
            1: Biscuit(1),
            2: Biscuit(2),
            3: Biscuit(3),
            4: Biscuit(4)
        }
        def find_valid_crossover_points(parent):
            valid_points = []
            for i in range(1, len(parent)):
                prev_position, prev_type = parent[i - 1]
                position, _ = parent[i]
                if position - (prev_position + biscuits[prev_type].length) > 0:
                    valid_points.append(i)
            return valid_points
    
        def create_child(p1, p2, crossover_point):
            child = p1[:crossover_point] + [b for b in p2[crossover_point:] if b not in p1[:crossover_point]]
            return child
    
        parent1 = sorted(parent1, key=lambda x: x[0])
        parent2 = sorted(parent2, key=lambda x: x[0])
    
        for _ in range(max_attempts):
            valid_points_parent1 = find_valid_crossover_points(parent1)
            valid_points_parent2 = find_valid_crossover_points(parent2)
    
            if valid_points_parent1 and valid_points_parent2:
                crossover_point1 = random.choice(valid_points_parent1)
                crossover_point2 = random.choice(valid_points_parent2)
    
                child1 = create_child(parent1, parent2, crossover_point1)
                child2 = create_child(parent2, parent1, crossover_point2)
    
                return child1, child2
    
        return parent1, parent2

    def evolve(self):
        
        elite_size = int(self.population_size * 0.2)
        selected = self.selection(elite_size)
        offspring = []
        
        new_population = selected[:elite_size]
    
        while len(new_population) < self.population_size:
            parent1 = selected[random.randint(0, elite_size - 1)]
            parent2 = selected[random.randint(0, elite_size - 1)]
            child1, child2 = self.uniform_crossover(parent1, parent2)
            new_population.append(child1)
            if len(new_population) < self.population_size:
                new_population.append(child2)
    
        self.population = new_population

In [117]:
def main():
    biscuits = {
        0: Biscuit(0),
        1: Biscuit(1),
        2: Biscuit(2),
        3: Biscuit(3),
        4: Biscuit(4)
    }
    dough = Dough(500)

    for _, row in defects_df.iterrows():
        dough.add_defect(position=row['x'], defect_class=row['class'])

    population_size = 150
    mutation_rate = 0.1
    crossover_rate = 0.5
    GA = UniformCrossoverGA(dough, biscuits, population_size, mutation_rate, crossover_rate)
    num_generations = 100
    
    for generation in range(num_generations):
        GA.evolve()
        print(f"Generation {generation + 1} complete.")

    best_solution = max(GA.population, key=GA.fitness)
    total_value = GA.fitness(best_solution)
    print("Best Solution Configuration:", best_solution)
    
    return best_solution

if __name__ == "__main__":
    best_solution = main()

Generation 1 complete.
Generation 2 complete.
Generation 3 complete.
Generation 4 complete.
Generation 5 complete.
Generation 6 complete.
Generation 7 complete.
Generation 8 complete.
Generation 9 complete.
Generation 10 complete.
Generation 11 complete.
Generation 12 complete.
Generation 13 complete.
Generation 14 complete.
Generation 15 complete.
Generation 16 complete.
Generation 17 complete.
Generation 18 complete.
Generation 19 complete.
Generation 20 complete.
Generation 21 complete.
Generation 22 complete.
Generation 23 complete.
Generation 24 complete.
Generation 25 complete.
Generation 26 complete.
Generation 27 complete.
Generation 28 complete.
Generation 29 complete.
Generation 30 complete.
Generation 31 complete.
Generation 32 complete.
Generation 33 complete.
Generation 34 complete.
Generation 35 complete.
Generation 36 complete.
Generation 37 complete.
Generation 38 complete.
Generation 39 complete.
Generation 40 complete.
Generation 41 complete.
Generation 42 complete.
G

In [118]:
biscuits_info = {
        0: Biscuit(0),
        1: Biscuit(1),
        2: Biscuit(2),
        3: Biscuit(3),
        4: Biscuit(4)
    }

In [119]:
dough = Dough(500)
for _, row in defects_df.iterrows():
    dough.add_defect(position=row['x'], defect_class=row['class'])
dough.place_biscuits_GA(best_solution)
print(dough)

Dough Defects:
Position: 355.45, Class: c
Position: 92.50, Class: a
Position: 141.88, Class: c
Position: 431.83, Class: c
Position: 435.03, Class: c
Position: 205.80, Class: a
Position: 34.69, Class: b
Position: 443.57, Class: a
Position: 69.42, Class: a
Position: 301.28, Class: a
Position: 342.65, Class: c
Position: 216.12, Class: a
Position: 367.67, Class: c
Position: 93.13, Class: a
Position: 401.42, Class: c
Position: 74.25, Class: c
Position: 356.84, Class: a
Position: 457.42, Class: a
Position: 395.64, Class: c
Position: 368.89, Class: a
Position: 163.92, Class: a
Position: 484.16, Class: b
Position: 55.27, Class: c
Position: 171.56, Class: b
Position: 119.69, Class: a
Position: 332.64, Class: b
Position: 85.10, Class: a
Position: 289.68, Class: b
Position: 156.79, Class: b
Position: 143.53, Class: c
Position: 83.61, Class: a
Position: 108.52, Class: a
Position: 57.58, Class: c
Position: 9.96, Class: c
Position: 440.00, Class: b
Position: 146.01, Class: b
Position: 259.33, Class:

In [120]:
dough.calculate_own_value(biscuits_info)

575

In [108]:
dough.validate_biscuits_placement(biscuits_info)

(True, 'All biscuits are correctly placed and within defect thresholds.')

<div style="background-color: #e8f4f8; border: 2px solid #b6d4e0; padding: 20px; border-radius: 10px; margin: 20px; box-shadow: 0 4px 6px rgba(0,0,0,0.1);">
    <h3 style="color: #2c7be5; font-size: 22px; font-family: 'Verdana', sans-serif; margin-bottom: 15px;">Observations:</h3>
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;">With our personalized set of parameters, our Genetic Algorithm goes up to results like 575 (and a little bit more on certain tries). This is pretty far from the optimum but considering the techniques used it is easy to find mistakes in patterns of the Genetic Algorithm. Now we will see how can a Greedy Search perform on this problem:
</p>

</div>

<a id="8"></a>
<h3 style="color: #4a86e8; font-size: 28px; font-family: 'Georgia', serif; text-shadow: 2px 2px 4px rgba(0,0,0,0.3); margin-top: 20px; border-left: 5px solid #4a86e8; padding-left: 10px;">Greedy Search</h3> 

<div style="background-color: #e8f4f8; border: 2px solid #b6d4e0; padding: 20px; border-radius: 10px; margin: 20px; box-shadow: 0 4px 6px rgba(0,0,0,0.1);">
    <h3 style="color: #2c7be5; font-size: 22px; font-family: 'Verdana', sans-serif; margin-bottom: 15px;">Introduction :</h3>
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;">A greedy search will function linearly to find a solution according to an heuristic decided. Here is our main code, changing its heursitic depending on parameters, and after a quick program allowing our code to launch and execute.
</p>

</div>

In [133]:
def greedy_biscuit_placement(dough, biscuits, heuristic = 'O'):
    
    if heuristic == 'value':
        sorted_biscuits = sorted(biscuits.values(), key=lambda b: b.value, reverse=True)
    elif heuristic == 'valength':
        sorted_biscuits = sorted(biscuits.values(), key=lambda b: b.value / b.length, reverse=True)
    else:
        sorted_biscuits = sorted(biscuits.values(), key=lambda b: b.value / (b.length*sum(b.max_defects.values())), reverse=True)
    
    total_value = 0
    current_position = 0
    placed_biscuits = []

    while current_position < dough.LENGTH:
        placed = False
        for biscuit in sorted_biscuits:
            # checks if biscuit fits in the remaining dough and meets defect requirements
            if current_position + biscuit.length <= dough.LENGTH:
                defects = dough.count_defects(current_position, biscuit.length)
                if all(defects.get(cls, 0) <= biscuit.max_defects[cls] for cls in biscuit.max_defects):
                    
                    placed_biscuits.append((current_position, biscuit.biscuit_type))
                    total_value += biscuit.value
                    current_position += biscuit.length
                    placed = True
                    break
            if not placed:
                current_position +=1

    return total_value, placed_biscuits

In [134]:
def main(heuristic):
    dough = Dough(length=500)
    for _, row in defects_df.iterrows():
        dough.add_defect(position=row['x'], defect_class=row['class'])
        
    biscuits = {
        0: Biscuit(0),
        1: Biscuit(1),
        2: Biscuit(2),
        3: Biscuit(3)
    }
    
    total_value, placed_biscuits = greedy_biscuit_placement(dough, biscuits,heuristic)
    print("Total Value:", total_value)
    print("Placed Biscuits:", placed_biscuits)

<a id="13"><h4 style="color: #6a737b; font-size: 24px; font-family: 'Courier New', monospace; margin-top: 15px; border-bottom: 1px dashed #6a737b; padding-bottom: 5px;">(i) First Heuristic: Value</h4></a>

In [135]:
if __name__ == "__main__":
    main('value')

Total Value: 658
Placed Biscuits: [(0, 1), (8, 1), (16, 1), (24, 1), (35, 2), (41, 1), (51, 0), (55, 1), (63, 1), (71, 1), (79, 1), (87, 1), (99, 1), (108, 3), (119, 0), (123, 1), (131, 1), (141, 0), (145, 1), (158, 3), (167, 1), (175, 1), (186, 2), (188, 1), (196, 1), (204, 1), (212, 1), (220, 1), (228, 1), (236, 1), (244, 1), (252, 1), (260, 1), (269, 3), (276, 0), (280, 1), (291, 2), (293, 1), (301, 1), (309, 1), (317, 1), (325, 1), (333, 1), (341, 1), (352, 2), (354, 1), (364, 0), (370, 0), (374, 1), (382, 1), (390, 1), (398, 1), (406, 1), (414, 1), (422, 1), (432, 0), (436, 1), (444, 1), (452, 1), (460, 1), (468, 1), (478, 0), (484, 0), (488, 1)]


<a id="14"><h4 style="color: #6a737b; font-size: 24px; font-family: 'Courier New', monospace; margin-top: 15px; border-bottom: 1px dashed #6a737b; padding-bottom: 5px;">(ii) Second Heuristic: Value / Length</h4></a>

In [136]:
if __name__ == "__main__":
    main('valength')

Total Value: 662
Placed Biscuits: [(1, 0), (6, 0), (10, 3), (15, 3), (21, 0), (26, 0), (30, 3), (36, 0), (41, 0), (45, 3), (51, 0), (56, 0), (60, 3), (65, 3), (71, 0), (76, 0), (82, 1), (91, 0), (99, 3), (104, 3), (114, 0), (119, 0), (124, 0), (128, 3), (133, 3), (139, 0), (144, 0), (148, 3), (158, 0), (167, 0), (172, 0), (177, 0), (186, 0), (190, 3), (195, 3), (200, 3), (205, 3), (210, 3), (215, 3), (220, 3), (225, 3), (230, 3), (235, 3), (242, 1), (250, 3), (255, 3), (260, 3), (265, 3), (270, 3), (276, 0), (281, 0), (285, 3), (294, 3), (300, 0), (305, 0), (309, 3), (315, 0), (319, 3), (324, 3), (330, 0), (334, 3), (339, 3), (344, 3), (349, 3), (356, 1), (365, 0), (370, 0), (375, 0), (383, 3), (388, 3), (393, 3), (399, 0), (403, 3), (408, 3), (413, 3), (418, 3), (424, 0), (429, 0), (434, 0), (439, 0), (444, 0), (448, 3), (454, 0), (458, 3), (463, 3), (468, 3), (473, 3), (479, 0), (484, 0), (488, 3), (493, 3)]


<a id="15"><h4 style="color: #6a737b; font-size: 24px; font-family: 'Courier New', monospace; margin-top: 15px; border-bottom: 1px dashed #6a737b; padding-bottom: 5px;">(iii) Third Heuristic: Value / (Length*Max number of defects)</h4></a>

In [139]:
if __name__ == "__main__":
    main('valengthdef')

Total Value: 674
Placed Biscuits: [(1, 0), (6, 0), (10, 3), (15, 3), (21, 0), (26, 0), (30, 3), (36, 0), (41, 0), (45, 3), (51, 0), (56, 0), (60, 3), (65, 3), (71, 0), (76, 0), (83, 1), (92, 0), (99, 1), (108, 0), (114, 2), (119, 1), (127, 3), (132, 3), (138, 0), (143, 0), (147, 3), (152, 3), (158, 0), (164, 2), (167, 0), (172, 0), (177, 0), (186, 0), (190, 3), (195, 3), (200, 3), (205, 3), (210, 3), (215, 3), (220, 3), (225, 3), (230, 3), (235, 3), (243, 1), (251, 3), (256, 3), (261, 3), (266, 3), (271, 3), (277, 0), (281, 3), (286, 3), (294, 1), (302, 3), (307, 3), (313, 0), (317, 3), (322, 3), (327, 3), (333, 0), (337, 3), (343, 0), (348, 0), (353, 0), (357, 3), (363, 0), (368, 0), (375, 1), (383, 3), (388, 3), (393, 3), (399, 0), (403, 3), (408, 3), (413, 3), (418, 3), (424, 0), (429, 0), (434, 0), (439, 0), (444, 0), (448, 3), (454, 0), (458, 3), (463, 3), (468, 3), (473, 3), (479, 0), (484, 0), (488, 3), (493, 3)]


<div style="background-color: #e8f4f8; border: 2px solid #b6d4e0; padding: 20px; border-radius: 10px; margin: 20px; box-shadow: 0 4px 6px rgba(0,0,0,0.1);">
    <h3 style="color: #2c7be5; font-size: 22px; font-family: 'Verdana', sans-serif; margin-bottom: 15px;">Observations:</h3>
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;">Our third heuristic is the best one so far, 674 being its result. These heuristics were easy to find but we could imagine to implement a more sophisticated one (linear variation of the reward due to placing or not successfully a type of biscuit, finding patterns in the defects, etc.).
</p>
</div>

<a id="16"><h4 style="color: #6a737b; font-size: 24px; font-family: 'Courier New', monospace; margin-top: 15px; border-bottom: 1px dashed #6a737b; padding-bottom: 5px;">(iv) Fourth Heuristic: without biscuit 2</h4></a>

<div style="background-color: #e8f4f8; border: 2px solid #b6d4e0; padding: 20px; border-radius: 10px; margin: 20px; box-shadow: 0 4px 6px rgba(0,0,0,0.1);">
    <h3 style="color: #2c7be5; font-size: 22px; font-family: 'Verdana', sans-serif; margin-bottom: 15px;">Introduction :</h3>
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;">As Biscuit 2 is our last biscuit in our heuristic order, we want to try to find out if any change happens if we remove it:
</p>

</div>

In [142]:
def main(heuristic):
    dough = Dough(length=500)
    for _, row in defects_df.iterrows():
        dough.add_defect(position=row['x'], defect_class=row['class'])
        
    biscuits = {
        1: Biscuit(1),
        0: Biscuit(0),
        3: Biscuit(3)
    }
    
    # Run Greedy Algorithm
    total_value, placed_biscuits = greedy_biscuit_placement(dough, biscuits,heuristic)
    print("Total Value:", total_value)
    print("Placed Biscuits:", placed_biscuits)

if __name__ == "__main__":
    main('ok')

Total Value: 674
Placed Biscuits: [(1, 0), (6, 0), (10, 3), (15, 3), (21, 0), (26, 0), (30, 3), (36, 0), (41, 0), (45, 3), (51, 0), (56, 0), (60, 3), (65, 3), (71, 0), (76, 0), (82, 1), (91, 0), (99, 0), (103, 3), (108, 3), (114, 0), (119, 0), (124, 0), (128, 3), (133, 3), (139, 0), (144, 0), (148, 3), (158, 1), (167, 0), (172, 0), (177, 0), (189, 1), (197, 3), (203, 0), (207, 3), (214, 1), (222, 3), (227, 3), (232, 3), (237, 3), (243, 0), (247, 3), (252, 3), (257, 3), (262, 3), (267, 3), (273, 0), (278, 0), (282, 3), (287, 3), (294, 1), (302, 3), (307, 3), (313, 0), (317, 3), (322, 3), (327, 3), (333, 0), (337, 3), (343, 0), (348, 0), (353, 0), (357, 3), (363, 0), (368, 0), (374, 1), (384, 1), (392, 3), (397, 3), (402, 3), (407, 3), (412, 3), (417, 3), (422, 3), (428, 0), (433, 0), (438, 0), (443, 0), (447, 3), (452, 3), (457, 3), (462, 3), (467, 3), (472, 3), (478, 0), (483, 0), (487, 3), (492, 3)]


<a id="9"></a>
<h3 style="color: #4a86e8; font-size: 28px; font-family: 'Georgia', serif; text-shadow: 2px 2px 4px rgba(0,0,0,0.3); margin-top: 20px; border-left: 5px solid #4a86e8; padding-left: 10px;">CSP with Ortools </h3> 

<div style="background-color: #e8f4f8; border: 2px solid #b6d4e0; padding: 20px; border-radius: 10px; margin: 20px; box-shadow: 0 4px 6px rgba(0,0,0,0.1);">
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;"><strong>Operations Research Tools</strong>, is an open-source software suite created by Google for solving combinatorial optimization problems. With this we can easily and fastly find good solutions or even optimum for CSP. Here is our implementation : 
</p>

</div>

<h4 style="color: #6a737b; font-size: 24px; font-family: 'Courier New', monospace; margin-top: 15px; border-bottom: 1px dashed #6a737b; padding-bottom: 5px;">(i) Import librairies</h4></a>


In [55]:
from ortools.sat.python import cp_model
model = cp_model.CpModel()

<h4 style="color: #6a737b; font-size: 24px; font-family: 'Courier New', monospace; margin-top: 15px; border-bottom: 1px dashed #6a737b; padding-bottom: 5px;">(ii) Create variables</h4></a>

In [56]:
#Variables
biscuit_types = [Biscuit(i) for i in range(4)]
for b in biscuit_types:
    print(b)

max_pos = dough.LENGTH - min(biscuit.length for biscuit in biscuit_types)#maximum position for the smallest biscuit
biscuit_vars = {}#We want to create a dictionnary that associate True if biscuit type b is placed at position p (b,p), False otherwise
for biscuit in biscuit_types:
    for start_pos in range(max_pos + 1):
        biscuit_vars[(biscuit.biscuit_type, start_pos)] = model.NewBoolVar(f'biscuit_{biscuit.biscuit_type}_start_{start_pos}')

Biscuit Type: 0, Length: 4, Value: 6, Max Defects: {'a': 4, 'b': 2, 'c': 3}
Biscuit Type: 1, Length: 8, Value: 12, Max Defects: {'a': 5, 'b': 4, 'c': 4}
Biscuit Type: 2, Length: 2, Value: 1, Max Defects: {'a': 1, 'b': 2, 'c': 1}
Biscuit Type: 3, Length: 5, Value: 8, Max Defects: {'a': 2, 'b': 3, 'c': 2}


In [57]:
print(biscuit_vars[(0,3)])

biscuit_0_start_3


<h4 style="color: #6a737b; font-size: 24px; font-family: 'Courier New', monospace; margin-top: 15px; border-bottom: 1px dashed #6a737b; padding-bottom: 5px;">(iii) Create Constraints</h4></a>

In [58]:
#No overlapping biscuits
for x in range(dough.LENGTH):
    # Initialize a constraint sum for each position x
    constraint_sum = 0

    for b in biscuit_types:
        for start_pos in range(max(0, x - b.length + 1), min(x + 1, max_pos + 1)):
            if x >= start_pos and x < start_pos + b.length:
                constraint_sum += biscuit_vars[(b.biscuit_type, start_pos)]

    # Add the constraint to the model
    model.Add(constraint_sum <= 1)


In [59]:
# Defects within acceptable limits

for biscuit in biscuit_types:
    for start_pos in range(max_pos + 1):
        for defect_type, max_allowed in biscuit.max_defects.items():
            # Initialize defect count for each biscuit type, start position, and defect type
            defect_count = 0

            # Count defects within the biscuit length
            for x, b in dough.defects_list:
                if b == defect_type and start_pos <= x < start_pos + biscuit.length:
                    defect_count += 1

            # Add the constraint to the model
            model.Add(defect_count * biscuit_vars[(biscuit.biscuit_type, start_pos)] <= max_allowed)

<h4 style="color: #6a737b; font-size: 24px; font-family: 'Courier New', monospace; margin-top: 15px; border-bottom: 1px dashed #6a737b; padding-bottom: 5px;">(iv) Objective function to maximize</h4></a>

In [60]:
 # Maximize the total value of the biscuits
total_value = sum(biscuit_vars[(biscuit.biscuit_type, start_pos)] * biscuit.value 
                      for biscuit in biscuit_types 
                      for start_pos in range(dough.LENGTH - biscuit.length + 1))#AND Length has to be in the 500 range ! 
model.Maximize(total_value)

<h4 style="color: #6a737b; font-size: 24px; font-family: 'Courier New', monospace; margin-top: 15px; border-bottom: 1px dashed #6a737b; padding-bottom: 5px;">(v) Solve the model</h4></a>

In [61]:
start =time.time()
solver = cp_model.CpSolver()
status = solver.Solve(model)
print("solved in : ", time.time()-start,"seconds")

solved in :  0.13903093338012695 seconds


In [62]:
if status in (cp_model.OPTIMAL, cp_model.FEASIBLE):
    if status==cp_model.OPTIMAL:
        print("The solution found is optimal !")
    if status==cp_model.FEASIBLE:
        print("The solution found is feasible !")
solution = []
for biscuit in biscuit_types:
    for start_pos in range(max_pos + 1):
        if solver.Value(biscuit_vars[(biscuit.biscuit_type, start_pos)]) == 1:#if 1/True this mean that it has been placed
            solution.append((biscuit.biscuit_type, start_pos))
print("The objective function value found is : ",solver.ObjectiveValue()," and the solution found is : ",solution)

The solution found is optimal !
The objective function value found is :  760.0  and the solution found is :  [(0, 9), (0, 36), (0, 40), (0, 50), (0, 54), (0, 58), (0, 75), (0, 84), (0, 93), (0, 104), (0, 114), (0, 118), (0, 122), (0, 136), (0, 140), (0, 144), (0, 148), (0, 158), (0, 162), (0, 167), (0, 171), (0, 175), (0, 186), (0, 200), (0, 214), (0, 218), (0, 242), (0, 246), (0, 275), (0, 279), (0, 314), (0, 336), (0, 345), (0, 362), (0, 366), (0, 370), (0, 374), (0, 393), (0, 430), (0, 434), (0, 438), (0, 442), (0, 446), (0, 475), (0, 479), (0, 483), (0, 487), (0, 496), (1, 0), (1, 23), (1, 67), (1, 293), (1, 301), (1, 328), (1, 354), (1, 422), (3, 13), (3, 18), (3, 31), (3, 45), (3, 62), (3, 79), (3, 88), (3, 99), (3, 108), (3, 126), (3, 131), (3, 152), (3, 179), (3, 190), (3, 195), (3, 204), (3, 209), (3, 222), (3, 227), (3, 232), (3, 237), (3, 250), (3, 255), (3, 260), (3, 265), (3, 270), (3, 283), (3, 288), (3, 309), (3, 318), (3, 323), (3, 340), (3, 349), (3, 378), (3, 383), (3

<div style="background-color: #e8f4f8; border: 2px solid #b6d4e0; padding: 20px; border-radius: 10px; margin: 20px; box-shadow: 0 4px 6px rgba(0,0,0,0.1);">
    <h3 style="color: #2c7be5; font-size: 22px; font-family: 'Verdana', sans-serif; margin-bottom: 15px;">Observations :</h3>
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;">Using the ortools librairy we suceed to find an optimal solution with a really good computation time, <strong>0.15 seconds on average</strong>. This demonstrate the power of python and it's librairies. OR-tools use constraint propagration and search algorithm like backtracking to improve the CSP speed.We need to note that despite the fact that it is slower than the dynamic programming approach, it is way faster than the genetic algorithm approach and it succeed to find an optimal solution. Futher more CSP approach are way more flexible, often problem cannot be solved by dynammic programming making the CSP approach with ortools one of the most convenient and fastest !
</p>

</div>

<a id="30"></a>
<div class='alert-block alert-info'>
    <br>
    <h1 align="center">Conclusion </h1>
    <br>
</div>

<div style="background-color: #e8f4f8; border: 2px solid #b6d4e0; padding: 20px; border-radius: 10px; margin: 20px; box-shadow: 0 4px 6px rgba(0,0,0,0.1);">
    <p style="line-height: 1.6; color: #333; font-family: 'Arial', sans-serif;">Comparing ourselves to the optimum, it is clear that the algorithms that we have implemented cannot reach it with the systems we used, but they can reach values close to it. It would require more complex programming and maybe some crossover of Algorithms. For this project, we know that the ideas we have developped still have potential and are content with our production. Thank you for reading until the end.
</p>

</div>

<div class='alert-block alert-info'>
    <br>
    <h1 align="center"><b>Thank you for reading!</b></h1>
    <h3 align="center">AI Algorithms - DIA 2</h3>
    <h5 align="center">Oscar Pastural - Capucine Boudin - Richard Goudelin</a></h5>
    <br>
</div>